pkg/syncer/pvcsi_fullsync.go

```diff
@@ -233,6 +241,121 @@ func createCnsVolumeMetadataList(ctx context.Context, metadataSyncer *metadataSy
 	return nil
 }
 
+// AddNodeAffinityRulesOnPV update TKG PVs with node affinity rules set on associated supervisor PVC if
+// node affinity rule is not set on the TKG PV
+func AddNodeAffinityRulesOnPV(ctx context.Context, metadataSyncer *metadataSyncInformer) {
+	log := logger.GetLogger(ctx)
+	log.Info("AddNodeAffinityRulesOnPV Start.")
+	pvList, err := getPVsInBoundAvailableOrReleased(ctx, metadataSyncer)
+	if err != nil {
+		log.Errorf("FullSync: Failed to get PVs from guest cluster. Err: %v", err)
+		return
+	}
+	// Get the supervisor namespace in which the guest cluster is deployed.
+	supervisorNamespace, err := cnsconfig.GetSupervisorNamespace(ctx)
+	if err != nil {
+		log.Errorf("FullSync: could not get supervisor namespace in which guest cluster was deployed. Err: %v", err)
+		return
+	}
+
+	// Create the kubernetes client from config.
+	k8sClient, err := k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("creating Kubernetes client failed. Err: %v", err)
+		return
+	}
+
+	for _, pv := range pvList {
+		if pv.Spec.NodeAffinity == nil {
+			supervisorPVClaim, err := metadataSyncer.supervisorClient.CoreV1().
+				PersistentVolumeClaims(supervisorNamespace).Get(ctx, pv.Spec.CSI.VolumeHandle, metav1.GetOptions{})
+			if err != nil {
+				log.Errorf("AddNodeAffinityRulesOnPV: failed to get supervisor PVC for the PV %v. Err: %v",
+					pv.Name, err)
+				continue
+			}
+			volumeAccessibleTopologyJSON := supervisorPVClaim.Annotations["csi.vsphere.volume-accessible-topology"]
+			var accessibleTopology []*csi.Topology
+			if volumeAccessibleTopologyJSON != "" {
+				accessibleTopology, err = parseJSONToTopologies(volumeAccessibleTopologyJSON)
+				if err != nil {
+					log.Errorf("failed to parse json %v to Topology. err :%v", volumeAccessibleTopologyJSON, err)
```

Comment:
nit: `failed to parse json %v to Topology. err :%v` -> `failed to parse json string %v to topology terms. Err :%v`

Reply:
replaced this function with `generateVolumeAccessibleTopologyFromPVCAnnotation`